File: tasks/2023/final/fire.py
#!/usr/bin/env python3
# @check-accepted: examples cubic quadratic
# @check-time-limit-exceeded: all

from sys import setrecursionlimit

setrecursionlimit(250_000)

N = int(input().strip())
G = [[] for _ in range(N)]
for _ in range(N-1):
    a, b = map(int, input().strip().split())
    G[a].append(b)
    G[b].append(a)

dp = [None] * N
ans = [None] * N


def cost(a):
    return sum(a) - max(a) + 1


def dfs1(v, p):
    dp[v] = [dfs1(u, v) for u in G[v] if u != p] + [0]
    dp[v].sort(reverse=True)
    return cost(dp[v])


def dfs2(v, p):
    s = sum(dp[v]) + 1
    ans[v] = s - dp[v][0]
    for u in G[v]:
        if u == p:
            continue
        c = cost(dp[u])
        s1 = s - c - dp[v][dp[v][0] == c]
        dp[u].append(s1)
        dp[u].sort(reverse=True)
        dfs2(u, v)
    return ans


dfs1(0, -1)
dfs2(0, -1)
for i in range(N):
    print(ans[i], end=' ')
print()

File: tasks/2023/final/fire.cpp
// @check-accepted: examples cubic quadratic all
#include <fstream>
#include <iostream>
#include <set>
#include <vector>
using namespace std;
vector<int> v[200005];
vector<int> ans(200005);
multiset<int> q[200005];
int sum[200005];
int ma[200005];
int dfs(int pos, int prec) {
    int m = 0;
    int tot = 0;
    for (int x : v[pos]) {
        if (x == prec) continue;
        int k = dfs(x, pos);
        tot += k;
        m = max(m, k);
        q[pos].insert(k);
    }
    sum[pos] = tot + 1;
    ma[pos] = m;
    q[pos].insert(0);
    q[pos].insert(0);
    return tot - m + 1;
}
void reroot(int pos, int prec) {
    ans[pos] = sum[pos] - ma[pos];

    for (int x : v[pos]) {
        if (x == prec) continue;
        sum[pos] -= (sum[x] - ma[x]);
        q[pos].erase(q[pos].find(sum[x] - ma[x]));
        ma[pos] = *prev(q[pos].end());
        sum[x] += sum[pos] - ma[pos];
        q[x].insert(sum[pos] - ma[pos]);
        ma[x] = *prev(q[x].end());

        reroot(x, pos);

        sum[x] -= sum[pos] - ma[pos];
        q[x].erase(q[x].find(sum[pos] - ma[pos]));
        ma[x] = *prev(q[x].end());
        sum[pos] += (sum[x] - ma[x]);
        q[pos].insert(sum[x] - ma[x]);
        ma[pos] = *prev(q[pos].end());
    }
}

int main() {
    ios::sync_with_stdio(false);
    int N;
    cin >> N;
    for (int i = 0, a, b; i < N - 1; i++) {
        cin >> a >> b;
        v[a].push_back(b);
        v[b].push_back(a);
    }
    dfs(0, -1);
    reroot(0, -1);
    for (int i = 0; i < N; i++) cout << ans[i] << " ";
    cout << '\n';
}
